plugins/pi4j-plugin-ffm/src/main/java/com/pi4j/plugin/ffm/providers/i2c/I2CSMBus.java

```diff
@@ -0,0 +1,149 @@
+package com.pi4j.plugin.ffm.providers.i2c;
+
+import com.pi4j.io.i2c.I2CBase;
+import com.pi4j.io.i2c.I2CConfig;
+import com.pi4j.io.i2c.I2CProvider;
+import com.pi4j.plugin.ffm.common.i2c.SMBusNative;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+public class I2CSMBus extends I2CBase<I2CBusFFM> {
+    private static final Logger logger = LoggerFactory.getLogger(I2CSMBus.class);
+    private static final SMBusNative SMBUS = new SMBusNative();
+
+    /**
+     * <p>Constructor for I2CBase.</p>
+     *
+     * @param provider a {@link I2CProvider} object.
+     * @param config   a {@link I2CConfig} object.
+     * @param i2CBus   a {@link I2CBus} object.
+     */
+    public I2CSMBus(I2CProvider provider, I2CConfig config, I2CBusFFM i2CBus) {
+        super(provider, config, i2CBus);
+    }
+
+
+    /**
+     * Internal method.
+     * Writes the data array into the register address of device selected previously.
+     *
+     * @param data data array to be written
+     */
+    private int writeInternal(int register, byte[] data) {
+        i2CBus.selectDevice(config.device());
+        return i2CBus.execute(this, (i2cFileDescriptor) -> {
+            if (data.length == 1) {
+                return SMBUS.writeByteData(i2cFileDescriptor, (byte) register, data[0]);
+            } else {
+                return 0;
+//                var buffer = new byte[data.length + 1];
+//                buffer[0] = (byte) register;
+//                System.arraycopy(data, 0, buffer, 1, data.length);
+//                return FILE.write(i2cFileDescriptor, buffer);
+            }
+        });
+    }
+
+    /**
+     * Internal method.
+     * Reads the data byte from the register address of device selected previously.
+     *
+     * @param register register address of selected device
+     * @return data byte read from register
+     */
+    private byte[] readInternal(int register, int size) {
+        i2CBus.selectDevice(config.device());
+        return i2CBus.execute(this, (i2cFileDescriptor) -> {
+            //FILE.write(i2cFileDescriptor, new byte[]{(byte) register});
+            //return FILE.read(i2cFileDescriptor, new byte[size], size);
+            if (size == 1) {
+                return new byte[] {SMBUS.readByteData(i2cFileDescriptor, (byte) register)};
+            } else {
+                return new byte[0];
+            }
+        });
+    }
+
+
+    @Override
+    public int read() {
+        i2CBus.selectDevice(config.device());
```

Comment:
Is this thread safe? One thread might select a device, then another selects another device and then the first does the execute...

Reply:
Indeed, not thread safety now :) Will add later on